dpnp/backend/src/queue_sycl.hpp

```diff
@@ -69,120 +60,95 @@ namespace mkl_rng = oneapi::mkl::rng;
+        auto &be = backend_sycl::get();
+        return *(be.queue_ptr);
     }
 
-public:
-    backend_sycl()
+    static mkl_rng::mt19937 &get_rng_engine()
     {
-#if defined(DPNP_LOCAL_QUEUE)
-        queue = nullptr;
-        rng_engine = nullptr;
-#endif
+        auto &be = backend_sycl::get();
+        return *(be.rng_mt19937_engine_ptr);
     }
 
-    virtual ~backend_sycl()
+    static mkl_rng::mcg59 &get_rng_mcg59_engine()
     {
-        backend_sycl::destroy();
+        auto &be = backend_sycl::get();
+        return *(be.rng_mcg59_engine_ptr);
     }
 
-    /**
-     * Explicitly disallow copying
-     */
-    backend_sycl(const backend_sycl &) = delete;
-    backend_sycl &operator=(const backend_sycl &) = delete;
-
-    /**
-     * Initialize @ref queue
-     */
-    static void backend_sycl_queue_init(
-        QueueOptions selector = QueueOptions::CPU_SELECTOR);
-
-    /**
-     * Return True if current @ref queue is related to cpu device
-     */
-    static bool backend_sycl_is_cpu();
-
-    /**
-     * Initialize @ref rng_engine and @ref rng_mcg59_engine
-     */
-    static void backend_sycl_rng_engine_init(size_t seed = 1);
-
-    /**
-     * Return the @ref queue to the user
-     */
-    static sycl::queue &get_queue()
+    template <typename SeedT>
+    void set_rng_engines_seed(const SeedT &seed)
     {
-#if defined(DPNP_LOCAL_QUEUE)
-        if (!queue) {
-            backend_sycl_queue_init();
+        auto rng_eng_mt19937 =
+            std::make_shared<mkl_rng::mt19937>(*queue_ptr, seed);
+        if (!rng_eng_mt19937) {
+            throw std::runtime_error(
+                "Could not create MT19937 engine with given seed");
         }
-
-        return *queue;
-#else
-        // temporal solution. Started from Sept-2020
-        DPCTLSyclQueueRef DPCtrl_queue = DPCTLQueueMgr_GetCurrentQueue();
-        if (DPCtrl_queue == nullptr) {
-            std::string reason =
-                (DPCTLQueueMgr_GetQueueStackSize() == static_cast<size_t>(-1))
-                    ? ": the queue stack is empty, probably no device is "
-                      "available."
-                    : ".";
+        auto rng_eng_mcg59 = std::make_shared<mkl_rng::mcg59>(*queue_ptr, seed);
+        if (!rng_eng_mcg59) {
             throw std::runtime_error(
-                "Failed to create a copy of SYCL queue with default device" +
-                reason);
+                "Could not create MCG59 engine with given seed");
         }
-        return *(reinterpret_cast<sycl::queue *>(DPCtrl_queue));
-#endif
+
+        rng_mt19937_engine_ptr.swap(rng_eng_mt19937);
+        rng_mcg59_engine_ptr.swap(rng_eng_mcg59);
     }
 
-    /**
-     * Return the @ref rng_engine to the user
-     */
-    static mkl_rng::mt19937 &get_rng_engine()
+    bool backend_sycl_is_cpu() const
     {
-        if (!rng_engine) {
-            backend_sycl_rng_engine_init();
-        }
-        return *rng_engine;
+        const sycl::queue &q = *queue_ptr;
```

Comment:
What is the intent of this line?

Reply:
Do you really intend to store a reference to the `sycl::queue`?

Reply:
It is just so that I can call `q.get_device()`. Of course `queue_ptr->get_device()` could be used directly.

Reply:
Resolved